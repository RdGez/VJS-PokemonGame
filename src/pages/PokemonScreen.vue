<template>
  <h1 class="fade-in">Who's That Pok√©mon?</h1>
  <div v-if="!pokemon">
    <br /><br /><br />
    <SpinnerPage />
  </div>
  <div v-else>
    <h4>{{ message }}</h4>
    <br />
    <PokemonPicture
      class="fade-in"
      :pokeId="pokemon.id"
      :showPokemon="showPokemon" />

    <PokemonOptions
      class="fade-in"
      @selection="checkAnswer($event)"
      :pokemons="arrPokemons"
      v-show="!showAnswer" />

    <button
      class="fade-in"
      @click="newGame"
      v-show="showAnswer">New
      Game</button>
  </div>
</template>

<script>
import PokemonPicture from "@/components/PokemonPicture";
import PokemonOptions from "@/components/PokemonOptions";
import SpinnerPage from "@/pages/SpinnerPage";

import getPokemonOptions from "@/helpers/getPokemonOptions";

export default {
  components: {
    PokemonPicture,
    PokemonOptions,
    SpinnerPage,
  },
  data() {
    return {
      message: null,
      pokemon: null,
      arrPokemons: [],
      showAnswer: false,
      showPokemon: false,
    };
  },
  methods: {
    async mixPokemonArray() {
      const rndInt = Math.floor(Math.random() * 4);

      this.arrPokemons = await getPokemonOptions();
      this.pokemon = this.arrPokemons[rndInt];
    },
    checkAnswer(pokemonId) {
      this.showAnswer = true;
      this.showPokemon = true;

      if (pokemonId == this.pokemon.id) {
        this.message = "Excellent! you guessed it";
      } else {
        this.message = `Oops! that pokemon is ${this.pokemon.name}`;
      }
    },
    newGame() {
      this.message = null;
      this.pokemon = null;
      this.arrPokemons = [];
      this.showAnswer = false;
      this.showPokemon = false;

      this.mixPokemonArray();
    }
  },
  mounted() {
    this.mixPokemonArray();
  },
};
</script>

<style scoped>
@import url("https://fonts.googleapis.com/css?family=Poppins:400,500,600,700&display=swap");
* {
  margin: 0;
  padding: 0;
  user-select: none;
  box-sizing: border-box;
  font-family: "Poppins", sans-serif;
}
button {
  background-color: #effbf2;
  border-radius: 2rem;
  border: 2px solid rgba(74, 150, 153, 100);
  cursor: pointer;
  width: 105px;
  margin-top: 40px;
  font-weight: 500;
  padding: 0.65rem 1rem;
  transition: 0.3s ease;
  margin-left: 10px;
}

button:hover {
  transition: 0.3s ease;
  background-color: #ade3cc;
}
</style>