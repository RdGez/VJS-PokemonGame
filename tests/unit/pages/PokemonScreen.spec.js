import { shallowMount } from "@vue/test-utils"
import PokemonScreen from "@/pages/PokemonScreen"

import { pokemons } from '../mocks/pokemons.mock';

describe("PokemonScreen Component", () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(PokemonScreen)
  })

  test("Debe hacer match con el snapshot", () => {
    expect( wrapper.html() ).toMatchSnapshot()
  })

  test('Debe de llamar el mixPokemonArray() al montar', () => {
    const mixPokemonArrSpy = jest.spyOn( PokemonScreen.methods, 'mixPokemonArray' )
    shallowMount(PokemonScreen)
    expect( mixPokemonArrSpy ).toHaveBeenCalled()
  })

  test('Debe de hacer match con el snapshot cuando cargan los pokemons', () => {
    const wrapper = shallowMount(PokemonScreen,{
      data() {
        return {
          arrPokemons: pokemons,
          message: null,
          pokemon: pokemons[0],
          showAnswer: false,
          showPokemon: false,
        }
      }
    })

    expect( wrapper.html() ).toMatchSnapshot()
  })

  test('Debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {
    const wrapper = shallowMount(PokemonScreen,{
      data() {
        return {
          arrPokemons: pokemons,
          message: null,
          pokemon: pokemons[0],
          showAnswer: false,
          showPokemon: false,
        }
      }
    })
    
    const picture = wrapper.find('pokemon-picture-stub')
    const options = wrapper.find('pokemon-options-stub')

    expect( picture.exists()).toBeTruthy()
    expect( options.exists()).toBeTruthy()
    expect( picture.attributes('pokeid')).toBe('1')
    expect( options.attributes('pokemons')).toBeTruthy()
  })  

  test('Pruebas con Checkanswer', async () => {
    const wrapper = shallowMount(PokemonScreen,{
      data() {
        return {
          arrPokemons: pokemons,
          message: null,
          pokemon: pokemons[0],
          showAnswer: false,
          showPokemon: false,
        }
      }
    })

    await wrapper.vm.checkAnswer(2)
    expect( wrapper.find('h4').exists() ).toBeTruthy()
    expect( wrapper.vm.showPokemon ).toBeTruthy()
  })
  
})
