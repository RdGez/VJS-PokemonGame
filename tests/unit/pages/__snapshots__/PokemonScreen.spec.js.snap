// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonScreen Component Debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<h1 class="fade-in">Who's That Pokémon?</h1>
<div>
  <h4></h4><br>
  <pokemon-picture-stub pokeid="1" showpokemon="false" class="fade-in"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]" class="fade-in"></pokemon-options-stub><button class="fade-in" style="display: none;">New Game</button>
</div>
`;

exports[`PokemonScreen Component Debe hacer match con el snapshot 1`] = `
<h1 class="fade-in">Who's That Pokémon?</h1>
<div><br><br><br>
  <spinner-page-stub></spinner-page-stub>
</div>
`;
